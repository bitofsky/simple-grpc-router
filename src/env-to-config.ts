export type ENV = {
    PATH_REDIRECT: string;
    NODE_ID?: string;
    NODE_CLUSTER?: string;
    ADMIN_PORT?: string;
    PORT?: string;
    OUTPUT_PATH?: string;
}

export type Route = {
    prefix: string;
    address: string;
    clusterName: string;
}

export type Cluster = {
    clusterName: string;
    host: string;
    port: number;
}

export type Config = {
    nodeId: string,
    nodeCluster: string,
    adminPort: number,
    listenPort: number,
    routeList: Route[];
    clusterMap: Map<string, Cluster>;
    outputPath: string;
}

export function envToConfig({
    PATH_REDIRECT,
    NODE_ID,
    NODE_CLUSTER,
    ADMIN_PORT,
    PORT,
    OUTPUT_PATH,
}: ENV): Config {

    const pathSet = new Set<string>();
    const clusterMap = new Map<string, Cluster>();
    const getClusterName = (address: string) => `cluster-${address.replace(/[^a-zA-Z0-9]/g, '-')}`;

    const routeList = (PATH_REDIRECT || '').split(',').map(s => {
        const [prefix, address] = s.split('->');

        if (!prefix || !address || !address.includes(':')) throw new Error(`Invalid route (${s})\nRoute must be in format 'PATH->HOST:PORT'`);

        if (pathSet.has(prefix)) throw new Error(`Duplicate path prefix (${prefix})`);

        pathSet.add(prefix);
        const clusterName = getClusterName(address);

        if (!clusterMap.has(clusterName)) {
            const [host, port] = address.split(':');
            clusterMap.set(clusterName, {
                clusterName,
                host,
                port: Number(port) || 50051,
            });
        }

        return { prefix, address, clusterName: getClusterName(address) };
    });

    return {
        nodeId: NODE_ID || 'envoy',
        nodeCluster: NODE_CLUSTER || 'envoy',
        adminPort: +(ADMIN_PORT || 9901),
        listenPort: +(PORT || 50051),
        clusterMap,
        routeList,
        outputPath: OUTPUT_PATH || '/tmp/config.yaml',
    };

}