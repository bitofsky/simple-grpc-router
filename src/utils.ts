import path from 'path';
import dotenv from 'dotenv';
import { debuglog } from 'util';

dotenv.config({ path: path.join(__dirname, '../.env') });

export const INFO = debuglog('INFO');
export const WARN = debuglog('WARN');
export const ERROR = debuglog('ERROR');

const ERROR_COMMON = `
Use 'export PATH_REDIRECT=<route>' to set it.

Example:
export PATH_REDIRECT=$PATH->$HOST:$PORT,$PATH->$HOST:$PORT,...
`.trim();

process.on('uncaughtException', (err) => {
    console.error(err.message + '\n\n' + ERROR_COMMON);
    process.exit(1);
});