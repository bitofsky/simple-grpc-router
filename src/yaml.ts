import type { EnvoyConfig } from "./interface";
import { dump } from 'js-yaml';

export function yamlDump(envoyConfig: EnvoyConfig): string {
    return dump(envoyConfig);
}
