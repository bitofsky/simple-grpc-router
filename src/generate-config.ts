import { INFO } from "./utils";
import type { Config } from "./env-to-config";
import type { Admin, Cluster, EnvoyConfig, FilterChain, Listener, Node, RouteElement, StaticResources } from "./interface";

export function generateConfig(cfg: Config): EnvoyConfig {

    if (!cfg.routeList?.length)
        throw new Error('No routes defined');

    INFO('Config', cfg);

    const node: Node = {
        id: cfg.nodeId,
        cluster: cfg.nodeCluster,
    };

    const admin: Admin = {
        access_log_path: '/tmp/admin_access.log',
        address: {
            socket_address: {
                address: '0.0.0.0',
                port_value: cfg.adminPort,
            },
        }
    };

    const routes: RouteElement[] = cfg.routeList.map(route => ({
        match: { path: route.path },
        route: {
            cluster: route.clusterName,
            timeout: '600s',
        },
    }));

    const listener: Listener = {
        name: 'listener',
        address: {
            socket_address: {
                address: '0.0.0.0',
                port_value: cfg.listenPort,
            },
        },
        filter_chains: [
            {
                filters: [
                    {
                        name: 'envoy.http_connection_manager',
                        typed_config: {
                            "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager",
                            access_log: [
                                {
                                    name: 'envoy.file_access_log',
                                    typed_config: {
                                        "@type": "type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog",
                                        path: '/dev/stdout',
                                    },
                                }
                            ],
                            stat_prefix: 'grpc_json',
                            codec_type: 'AUTO',
                            route_config: {
                                name: 'local_route',
                                virtual_hosts: [
                                    {
                                        name: 'local_service',
                                        domains: ['*'],
                                        routes: [
                                            ...routes,
                                            {
                                                match: { path: '/' },
                                                direct_response: {
                                                    status: 200,
                                                    body: {
                                                        inline_string: 'ok',
                                                    }
                                                }
                                            },
                                        ],
                                        retry_policy: {
                                            retry_on: 'connect-failure,refused-stream,unavailable',
                                            num_retries: 3,
                                            per_try_timeout: '3600s',
                                            retry_back_off: {
                                                base_interval: '2s',
                                                max_interval: '2s',
                                            },
                                        },
                                    }
                                ],
                            },
                            upgrade_configs: { upgrade_type: 'websocket', enabled: true },
                            http_filters: [{
                                name: 'envoy.filters.http.grpc_web',
                                typed_config: { "@type": 'type.googleapis.com/envoy.extensions.filters.http.grpc_web.v3.GrpcWeb' }
                            }, {
                                name: 'envoy.filters.http.cors',
                                typed_config: { "@type": 'type.googleapis.com/envoy.extensions.filters.http.cors.v3.Cors' }
                            }, {
                                name: 'envoy.filters.http.router',
                                typed_config: { "@type": 'type.googleapis.com/envoy.extensions.filters.http.router.v3.Router' }
                            }]
                        }
                    }
                ],
            },
        ],
    };

    const listeners: Listener[] = [listener];
    const clusters: Cluster[] = [...cfg.clusterMap.values()].map(cluster => ({
        name: cluster.clusterName,
        connect_timeout: '1.25s',
        type: 'strict_dns',
        lb_policy: 'round_robin',
        dns_lookup_family: 'V4_ONLY',
        load_assignment: {
            cluster_name: cluster.clusterName,
            endpoints: [{
                lb_endpoints: [{
                    endpoint: {
                        address: {
                            socket_address: {
                                address: cluster.host,
                                port_value: cluster.port,
                            },
                        },
                    }
                }]
            }]
        },
        http2_protocol_options: {},
        circuit_breakers: {
            thresholds: {
                max_pending_requests: 100000000,
                max_requests: 100000000,
            },
        },
    }));

    const static_resources: StaticResources = {
        listeners,
        clusters,
    }

    return {
        node,
        admin,
        static_resources,
    };
}
