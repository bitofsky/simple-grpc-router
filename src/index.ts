import { ENV, envToConfig } from "./env-to-config";
import { generateConfig } from "./generate-config";
import { yamlDump } from "./yaml";
import fs from "fs";

console.log('Generating config...');

const config = envToConfig(process.env as ENV);
const yaml = yamlDump(generateConfig(config));

fs.writeFileSync(config.outputPath, yaml);

console.log(`\n${yaml}`);
console.log(`Wrote config to ${config.outputPath}`);

process.exit(0);
