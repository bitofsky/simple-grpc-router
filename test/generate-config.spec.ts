import { ENV, envToConfig } from '../src/env-to-config';
import { generateConfig } from '../src/generate-config';

type TEST = {
    name: string;
    env: ENV;
}

const defaultEnv = {
};

const tests: TEST[] = [
    {
        name: 'basic',
        env: {
            ...defaultEnv,
            PATH_REDIRECT: '/->localhost:50051,/test1/->localhost:50051,/test2/->localhost:50052',
        }
    }
];

describe('generate', () => {
    for (const test of tests) {
        it(test.name, () => {
            const config = envToConfig(test.env);
            expect(generateConfig(config)).toMatchSnapshot();
        });
    }
});
