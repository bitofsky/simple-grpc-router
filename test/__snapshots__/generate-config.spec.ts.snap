// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate basic 1`] = `
Object {
  "admin": Object {
    "access_log_path": "/tmp/admin_access.log",
    "address": Object {
      "socket_address": Object {
        "address": "0.0.0.0",
        "port_value": 9901,
      },
    },
  },
  "node": Object {
    "cluster": "envoy",
    "id": "envoy",
  },
  "static_resources": Object {
    "clusters": Array [
      Object {
        "circuit_breakers": Object {
          "thresholds": Object {
            "max_pending_requests": 100000000,
            "max_requests": 100000000,
          },
        },
        "connect_timeout": "1.25s",
        "dns_lookup_family": "V4_ONLY",
        "http2_protocol_options": Object {},
        "lb_policy": "round_robin",
        "load_assignment": Object {
          "cluster_name": "cluster-localhost-50051",
          "endpoints": Array [
            Object {
              "lb_endpoints": Array [
                Object {
                  "endpoint": Object {
                    "address": Object {
                      "socket_address": Object {
                        "address": "localhost",
                        "port_value": 50051,
                      },
                    },
                  },
                },
              ],
            },
          ],
        },
        "name": "cluster-localhost-50051",
        "type": "strict_dns",
      },
      Object {
        "circuit_breakers": Object {
          "thresholds": Object {
            "max_pending_requests": 100000000,
            "max_requests": 100000000,
          },
        },
        "connect_timeout": "1.25s",
        "dns_lookup_family": "V4_ONLY",
        "http2_protocol_options": Object {},
        "lb_policy": "round_robin",
        "load_assignment": Object {
          "cluster_name": "cluster-localhost-50052",
          "endpoints": Array [
            Object {
              "lb_endpoints": Array [
                Object {
                  "endpoint": Object {
                    "address": Object {
                      "socket_address": Object {
                        "address": "localhost",
                        "port_value": 50052,
                      },
                    },
                  },
                },
              ],
            },
          ],
        },
        "name": "cluster-localhost-50052",
        "type": "strict_dns",
      },
    ],
    "listeners": Array [
      Object {
        "address": Object {
          "socket_address": Object {
            "address": "0.0.0.0",
            "port_value": 50051,
          },
        },
        "filter_chains": Array [
          Object {
            "filters": Array [
              Object {
                "name": "envoy.http_connection_manager",
                "typed_config": Object {
                  "@type": "type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager",
                  "access_log": Array [
                    Object {
                      "name": "envoy.file_access_log",
                      "typed_config": Object {
                        "@type": "type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog",
                        "path": "/dev/stdout",
                      },
                    },
                  ],
                  "codec_type": "AUTO",
                  "http_filters": Array [
                    Object {
                      "name": "envoy.filters.http.grpc_web",
                      "typed_config": Object {
                        "@type": "type.googleapis.com/envoy.extensions.filters.http.grpc_web.v3.GrpcWeb",
                      },
                    },
                    Object {
                      "name": "envoy.filters.http.cors",
                      "typed_config": Object {
                        "@type": "type.googleapis.com/envoy.extensions.filters.http.cors.v3.Cors",
                      },
                    },
                    Object {
                      "name": "envoy.filters.http.router",
                      "typed_config": Object {
                        "@type": "type.googleapis.com/envoy.extensions.filters.http.router.v3.Router",
                      },
                    },
                  ],
                  "route_config": Object {
                    "name": "local_route",
                    "virtual_hosts": Array [
                      Object {
                        "domains": Array [
                          "*",
                        ],
                        "name": "local_service",
                        "retry_policy": Object {
                          "num_retries": 3,
                          "per_try_timeout": "3600s",
                          "retry_back_off": Object {
                            "base_interval": "2s",
                            "max_interval": "2s",
                          },
                          "retry_on": "connect-failure,refused-stream,unavailable",
                        },
                        "routes": Array [
                          Object {
                            "match": Object {
                              "prefix": "/",
                            },
                            "route": Object {
                              "cluster": "cluster-localhost-50051",
                              "timeout": "600s",
                            },
                          },
                          Object {
                            "match": Object {
                              "prefix": "/test1/",
                            },
                            "route": Object {
                              "cluster": "cluster-localhost-50051",
                              "timeout": "600s",
                            },
                          },
                          Object {
                            "match": Object {
                              "prefix": "/test2/",
                            },
                            "route": Object {
                              "cluster": "cluster-localhost-50052",
                              "timeout": "600s",
                            },
                          },
                          Object {
                            "direct_response": Object {
                              "body": Object {
                                "inline_string": "ok",
                              },
                              "status": 200,
                            },
                            "match": Object {
                              "path": "/",
                            },
                          },
                        ],
                      },
                    ],
                  },
                  "stat_prefix": "grpc_json",
                  "upgrade_configs": Object {
                    "enabled": true,
                    "upgrade_type": "websocket",
                  },
                },
              },
            ],
          },
        ],
        "name": "listener",
      },
    ],
  },
}
`;
